### PR TITLE
Download TempIds on Wifi connections

### PR DESCRIPTION
**Pull request by fustilio**
Quite a big change across many files as the BLE had to be deinitialized before performing https request and re-initalize after the request is done. 

Added many debug statements as the feature is still a work-in-progress until we implement other stuff.

In the setup function, if no settings are found, some default settings are written to the device.

### CODE REVIEW

File: alpha/alpha.ino [hunk 1]
```diff
@@ -20,6 +25,36 @@ void setup() {
...
+
+  log_i("Stored settings:");
+  log_i("Version: %d", settings->settingsVersion);
+  log_i("UID: %.32s", settings->userId);
+  log_i("WIFI_SSID: %.32s", settings->wifiSsid);
+  log_i("WIFI_PASS: %.32s", settings->wifiPass);
+
+  TS_Storage.settings_save();
+
+  TS_RADIO.config(settings->wifiSsid, settings->wifiPass, settings->userId);
```
Comment (jamesi8086) [inline alpha/alpha.ino]: TS_RADIO can actually access TS_Storage directly

File: alpha/radio.cpp [hunk 2]
```diff
@@ -1,25 +1,56 @@
...
 #define SSID_DISPLAY_COUNT 1
 
-#define WIFI_RETRY_INTERVAL 30000
+#define WIFI_RETRY_INTERVAL 5000
+
+#define TEMPID_BATCH_SIZE 100
+#define HTTPS_PORT 443
+const char* host = "asia-east2-bettr-trace-1.cloudfunctions.net";
+const char* uri_getTempIDs = "/getTempIDs";
+const char* root_ca = \
```
Comment (jamesi8086) [inline alpha/radio.cpp]: can you use F(..) strings so that these are in flash and not RAM ?

File: alpha/radio.cpp [hunk 3]
```diff
@@ -99,7 +139,118 @@ void _TS_RADIO::wifi_update()
...
+    }
+
+    // Skip HTTP headers
+    if (!client->find(endOfHeaders)) {
+      log_e("Invalid response");
+      client->stop();
+      return;
+    }
+    
+    const size_t capacity = JSON_ARRAY_SIZE(TEMPID_BATCH_SIZE) + JSON_OBJECT_SIZE(1) + (TEMPID_BATCH_SIZE + 1)*JSON_OBJECT_SIZE(3) + 12610;
```
Comment (jamesi8086) [inline alpha/radio.cpp:206]: this would be really large, and thus require 2n space

File: alpha/radio.cpp (hunk 2 above)
Comment (fustilio) [reply alpha/radio.cpp]: F doesn't seem to work, I'll use PROGMEM instead

File: alpha/radio.h [hunk 4]
```diff
@@ -13,19 +11,25 @@ class _TS_RADIO
...
   private:
     void wifi_connect();
     void wifi_disconnect();
     bool wifi_scan_networks();
+    void download_temp_ids();
 
     bool wifiEnabled;
     long wifiTimerStart;
+
+    char ssid[32];
```
Comment (jamesi8086) [inline alpha/radio.h]: we probably don't need ssid, password here if this can get from TS_Storage

File: alpha/radio.cpp [hunk 5]
```diff
@@ -77,29 +114,147 @@ void _TS_RADIO::wifi_enable(bool enable)
...
+    if (!client->connect(host, HTTPS_PORT)) 
+    { 
+      log_e("Connection to %s failed", host);
+      return;
+    }
+
+    char body[52];
+    sprintf(body, "{\"data\":{\"uid\":\"%s\"}}", this->uid);
+
+    String postRequest =
```
Comment (jamesi8086) [inline alpha/radio.cpp]: can we use std::string ?

one of the guidelines is to not use arduino's String class, which supposedly sucks at memory usage

File: alpha/radio.cpp [hunk 6]
```diff
@@ -77,29 +114,147 @@ void _TS_RADIO::wifi_enable(bool enable)
...
+    { 
+      log_e("Connection to %s failed", host);
+      return;
+    }
+
+    char body[52];
+    sprintf(body, "{\"data\":{\"uid\":\"%s\"}}", this->uid);
+
+    String postRequest = 
+      String("POST ") + uri_getTempIDs + " HTTP/1.1\r\n" +
```
Comment (jamesi8086) [inline alpha/radio.cpp]: this part potentially can use a PROGMEM string with placeholders e.g. %s
excessive string concat with + is expensive, supposedly does a string copy of 2 halves to form new string.

File: alpha/radio.cpp [hunk 7]
```diff
@@ -77,29 +114,147 @@ void _TS_RADIO::wifi_enable(bool enable)
...
+      log_e("Failed to send request");
+      client->stop();
+      return;
+    }
+
+    // Check HTTP status
+    char status[32] = {0};
+    client->readBytesUntil('\r', status, sizeof(status));
+    if (strcmp(status, "HTTP/1.1 200 OK") != 0) {
+      log_e("Unexpected response %s", status);
```
Comment (jamesi8086) [inline alpha/radio.cpp:194]: log_w for non-fatal, acceptable errors

File: alpha/radio.cpp [hunk 8]
```diff
@@ -1,25 +1,62 @@
...
 
-#define WIFI_RETRY_INTERVAL 30000
+#define WIFI_RETRY_INTERVAL 5000
+
+#define TEMPID_BATCH_SIZE 100
+#define HTTPS_PORT 443
+const PROGMEM char* host = "asia-east2-bettr-trace-1.cloudfunctions.net";
+const PROGMEM char* uri_getTempIDs = "/getTempIDs";
+const PROGMEM char* endOfHeaders = "\r\n\r\n";
+const PROGMEM char* root_ca = \
```
Comment (jamesi8086) [inline alpha/radio.cpp:29]: woohoo

File: alpha/radio.cpp [hunk 9]
```diff
@@ -1,25 +1,62 @@
...
-#define WIFI_PASS "password"    // Enter your WiFi password here
-#endif
 #define SSID_DISPLAY_COUNT 1
 
-#define WIFI_RETRY_INTERVAL 30000
+#define WIFI_RETRY_INTERVAL 5000
+
+#define TEMPID_BATCH_SIZE 100
+#define HTTPS_PORT 443
+const PROGMEM char* host = "asia-east2-bettr-trace-1.cloudfunctions.net";
```
Comment (jamesi8086) [inline alpha/radio.cpp:18]: I wonder if this can go into settings as default

**Review by jamesi8086 — APPROVED** [review]
w/ improvements

**Pull request merged**